### ISSUE TITLE
Add dtype-aware numerical precision info (finfo equivalent)

### ISSUE THREAD

**Issue by antimora** [tracel-ai/burn#4700]
## Summary

Burn lacks a way to query numerical precision constants (epsilon, min, max, smallest normal) for a given float dtype. This forces code to hardcode magic constants for numerical stability, which may not be appropriate across all precisions (f16, bf16, f32, f64).

This is needed at two levels:
1. **Backend/tensor level** - for ops like `attention_fallback` that work with `FloatTensor<B>`
2. **Element/kernel level** - for CubeCL kernels and similar code that operates on individual elements of type `E` and needs `E::epsilon()` or `E::tiny()` directly

## Motivation

In #4694 (PR #4697), fixing NaN in `attention_fallback` required hardcoded clamp constants (`-1e4`, `1e-6`). These work for f32 but are not derived from dtype properties. Other places in the codebase also hardcode precision-dependent values:

- All norm modules use `epsilon = 1e-5`
- Cosine similarity uses `DEFAULT_EPSILON = 1e-8`
- Softmax activation uses max-normalization without a dtype-aware floor

CubeCL kernels face the same problem at the element level. For example, a `recip_inplace` that needs to safely handle zero-valued rows (from fully-masked attention) must hardcode a threshold constant:

```rust
pub fn recip_inplace(&mut self) {
    for i in 0..self.num_rows {
        let row_val = self.vals[i];
        // Hardcoded threshold - should be E::tiny() or E::epsilon()
        let epsilon = E::new(FULLY_MASKED_ROW_THRESHOLD);
        let not_masked = E::cast_from(row_val >= epsilon);
        let safe_val = clamp_min(row_val, epsilon);
        let recip = safe_val.recip();
        self.vals[i] = not_masked * recip;
    }
}
```

With dtype-aware constants, this would use `E::tiny()` or `E::epsilon()` instead of a hardcoded value, automatically adapting to f16/bf16/f32/f64.

## Prior art

All major ML frameworks expose this via a `finfo(dtype)` API modeled after NumPy:

- **PyTorch**: `torch.finfo(dtype).eps`, `.min`, `.max`, `.tiny` (smallest normal)
- **JAX**: `jnp.finfo(dtype)` (same fields)
- **TensorFlow**: `tf.experimental.numpy.finfo(dtype)`

Common usage patterns:
```python
# Masked fill with dtype-aware minimum
attn = attn.masked_fill(mask, torch.finfo(attn.dtype).min)

# Dtype-aware epsilon for numerical stability
norm = x / (x.std(dim=-1, keepdim=True) + torch.finfo(x.dtype).eps)
```

### Element level (for kernels)

Add associated constants or methods to `FloatElement` / CubeCL's numeric element trait:

```rust
trait FloatElement: Element {
    /// Machine epsilon: smallest value where 1.0 + eps != 1.0
    const EPSILON: Self;
    /// Largest representable finite value
    const MAX: Self;
    /// Most negative representable finite value
    const MIN: Self;
    /// Smallest positive normal value
    const TINY: Self;
}
```

This lets kernel code write `E::EPSILON` or `E::TINY` and get the right value for whatever precision the kernel is instantiated with.

### Backend/tensor level

Add a `FloatDTypeInfo` struct accessible from `FloatDType` or `Backend`:

```rust
pub struct FloatDTypeInfo {
    pub eps: f64,
    pub max: f64,
    pub min: f64,
    pub tiny: f64,
}
```

This could be queryable via `DeviceSettings` or a method on `FloatDType`. The exact API design is open for discussion.

## Impact

Would allow numerical stability code across the codebase (attention, normalization, activation functions, CubeCL kernels) to use dtype-appropriate constants instead of hardcoded values, automatically adapting to different precisions.

**Comment by antimora** [comment]
@louisfd @laggui @nathanielsimard

**Comment by laggui** [comment]
Bit of a duplicate, but linking to https://github.com/tracel-ai/burn/issues/4367
